src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+                    * (3.0 * u_f64 * d / 4.0)
+                    + d
+                    + u_f64)
+                    .floor() as usize,
+            }
+        } else {
+            // Case 2, Theorem 13, ie λ^3 > n_p > λ^2
+            let lambda_rel1 = lambda_rel.min(s1);
+            let lbar = (lambda_rel1 + 7.0) / loge;
+            let d = (16.0 * u_f64 * lbar).ceil();
+            assert!(n_p_f64 >= d * d / (9.0 * lbar));
+
+            let w = compute_w(u_f64, lambda_rel1);
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel1).ceil() as usize,
+                d: d as usize,
+                q: (2.0 * lbar / d).recip(),
+                b: (((w * lbar) / d + 1.0)
+                    * E.powf(2.0 * u_f64 * w * lbar / n_p_f64 + 7.0 * u_f64 / w)
+                    * d
+                    * u_f64
+                    + d)
+                    .floor() as usize,
+            }
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    v: usize,
+    /// Proof 2nd counter
+    t: usize,
+    // Round candidate tuple
+    s_list: Vec<Data>,
+    /// Round candidate hash
+    h: Hash,
+    /// Round candidate hash mapped to [1, n_p]
+    h_usize: usize,
+    /// Approximate size of set Sp to lower bound
+    n_p: usize,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(data: Vec<Vec<u8>>, n_p: usize) -> (Hash, usize) {
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        (digest, utils::oracle_uniform(&digest, n_p))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: usize, t: usize, n_p: usize) -> Round {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(t.to_ne_bytes().to_vec());
+        let (h, h_usize) = Round::h1(data, n_p);
+        Round {
+            v,
+            t,
+            s_list: vec![],
+            h,
+            h_usize,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: Data) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = vec![r.h.clone().to_vec()];
+        data.push(s.to_vec());
+        let (h, h_usize) = Round::h1(data, r.n_p);
+        Round {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h,
+            h_usize,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    r: usize,
```

Comment:
We should call the variables `v`, `t`, `elements` to be consistent with the paper.

Reply:
@tolikzinovyev , @rrtoledo : How about we give them a more descriptive namings in the code and explain in comments how they map to the notation in the paper?

Reply:
Renamed v,t, for the time being.
Opened new issue for more descriptive names

Reply:
Moved to new issue https://github.com/cardano-scaling/alba/issues/38